### DISCUSSION TITLE
useObservable deps being negated by preloadObservable

### DISCUSSION THREAD

**Discussion by italodeandra** [FirebaseExtended/reactfire#256] [Q&A]
I was trying to use the `useObservable` with the **deps** property. Inside the `useObservable` code, the useEffect works fine when the *deps* is changed. But as we filled with an **observableId**, the `preloadObservable` negates the changes on the `useObservable` **deps** because it is using a cached observable that matches the **observableId** filled.

I need help with that because I want the `useObservable` to re-subscribe using the new values I passed to the observable.

References:
https://github.com/FirebaseExtended/reactfire/blob/68cfc7957caa317322fc745ad0880b34a4df1065/reactfire/useObservable/index.ts#L45
https://github.com/FirebaseExtended/reactfire/blob/68cfc7957caa317322fc745ad0880b34a4df1065/reactfire/useObservable/index.ts#L19

**Comment by mrmckeb** [comment]
I've been playing with this a little and I feel that this may be causing #228.

Internally, the observable is created with a document ID (as an example) and when a user logs in, then logs out, then logs in - the old observable is used again and throws an error.

Perhaps there needs to be cleanup on `preloadedObservables`?

**Comment by italodeandra** [comment]
> I've been playing with this a little and I feel that this may be causing #228.
> 
> Internally, the observable is created with a document ID (as an example) and when a user logs in, then logs out, then logs in - the old observable is used again and throws an error.
> 
> Perhaps there needs to be cleanup on `preloadedObservables`?

I'll take a look in your case after work. But I think what they do with "caching" the observable is maybe a good thing.
In my case I fixed by changing the ID when I change the variable inside the observable. And I think it's good because if the user changes the variable back to a observable that already has an ID, it will reuse the cached observable (so the fetch goes way faster the second time).
But as you said, if it throws an error so there might also be another problem happening for you. I'll take a look in your case after I finish working, let me know if you fix it differently.

**Comment by mrmckeb** [comment]
I definitely agree that caching makes sense - the issue is more that cached observable fails if the user logs out and then back in. Right now I'm manually cleaning up the cached observables to try to solve that.
